AEM Cloud Configuration

hlbraddock opened the pull request:
This PR provides changes in support of proxying requests to the AEM Cloud CDN.

The proxy server effectively acts as "CDN" for www.cru.org.  See CloudFront CDN configuration 'companion' PR https://github.com/CruGlobal/cru-terraform/pull/2473 for non www.cru.org site(s).

These changes are (in theory) backwards compatible with AEM on premise solution.

Target **staging** delivery date 8/1.
Target prod-cloud.cru.org **production** delivery date 1/10/23.
Target cru.org **production** delivery date 3/31/23.

Note that the proxy server now also needs to be able to handle proxying SSL requests to the AEM Adobe Cloud CDN (i.e. `DEFAULT_PROXY_TARGET` https protocol). (not sure if it currently can do so).

[CDN in AEM as a Cloud Service](https://experienceleague.adobe.com/docs/experience-manager-cloud-service/content/implementing/content-delivery/cdn.html):

Note that this PR does not set `Host` header as per instructions below (nor does the companion PR referenced above as CloudFront prohibits setting the header).  However, it may very well be that it need not be set as per instructions below (and in fact possibly should not as Adobe documentation has before been found to be incorrect). We'll verify during testing in staging.

> **Configuration instructions:**
> 
> Point your CDN to the Adobe CDN’s ingress as its origin domain. For example, publish-p<PROGRAM_ID>-e<ENV-ID>.adobeaemcloud.com.
> 
> SNI must also be set to the Adobe CDN’s ingress.
> 
> Set the Host header to the origin domain. For example: Host:publish-p<PROGRAM_ID>-e<ENV-ID>.adobeaemcloud.com.
> 
> Set the X-Forwarded-Host header with the domain name so AEM can determine the host header. For example: X-Forwarded-Host:example.com.
> 
> Set X-AEM-Edge-Key. The value should come from Adobe.

cru-Luis-Rodriguez reviewed (approved):
That looks good to me

Omicron7 commented on main/location.conf.template:
```diff
@@ -58,6 +58,8 @@ location / {
   proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
   proxy_set_header X-Forwarded-Proto $thescheme;
+  proxy_set_header X-Forwarded-Host $host;
+  proxy_set_header X-AEM-Edge-Key $AEM_EDGE_KEY
```
Any ENV var used in nginx/lua must also be defined as an `ENV` key. This means adding `AEM_EDGE_KEY` to the [env-site.conf](https://github.com/CruGlobal/cruorg_proxy/blob/271a1d432b555380e83163cbc9907a1f33f7840c/main/env-site.conf) file.

Omicron7 replied:
Docs https://nginx.org/en/docs/ngx_core_module.html#env

hlbraddock commented:
@Omicron7 We'd like to deploy this in staging and believe it should be backwards compatible with production solution. If you agree, would you approve?

Omicron7 commented:
> @Omicron7 We'd like to deploy this in staging and believe it should be backwards compatible with production solution. If you agree, would you approve?

@hlbraddock I agree. Feel free to merge this to the `staging` branch where it'll get built and deployed to stage.
Also, make sure `AEM_EDGE_KEY` exists in stage secrets before building since the build bakes it into the image.

hlbraddock commented:
Done.
> @hlbraddock I agree. Feel free to merge this to the `staging` branch where it'll get built and deployed to stage. Also, make sure `AEM_EDGE_KEY` exists in stage secrets before building since the build bakes it into the image.

Any objections to merging this PR to `master` as well? @Omicron7

hlbraddock commented:
Any objections to merging this PR to master as well? @Omicron7

Omicron7 commented:
Assuming you've tested it on `staging` and added the `AEM_EDGE_KEY`, it should be fine. If it isn't needed right away, I'd suggest waiting til after Cru22. @hlbraddock

cru-Luis-Rodriguez committed:
Set AEM_ADDR as the Proxy `Host` value.
Per Adobe CDN documentation.
https://experienceleague.adobe.com/docs/experience-manager-cloud-service/content/implementing/content-delivery/cdn.html?lang=en
```
Set the Host header to the origin domain. For example: Host:publish-p<PROGRAM_ID>-e<ENV-ID>.adobeaemcloud.com.
```

Omicron7 commented:
@hlbraddock I'm going to be re-working the Docker build for this app for ECS multi-container work. This OK to merge to production?

hlbraddock commented:
I'm not quite sure of impact for cruorg proxy when **forwarding on to AEM premise**, particularly the headers being set [here](https://github.com/CruGlobal/cruorg_proxy/pull/27/files#diff-a5c89492c43628e913541619fb9ea7ec0a195b3017556d1f62140c4b5e4d30be).
(and we haven't yet configured `AEM_ADDR` env vars for staging/production, though we have `AEM_EDGE_KEY`, which would be irrelevant AEM on premise).

We've only tested this PR with cruorg proxy forwarding to the AEM cloud, not forwarding to AEM on premise.

It was only intended to be merged once we migrated to AEM Cloud and were then forwarding requests to the AEM Cloud.

I wouldn't feel confident merging during AEM on premise forwarding, until I did further analysis.

cc: @cru-Luis-Rodriguez

Omicron7 commented on main/location.conf.template:
```diff
@@ -54,10 +54,12 @@ location / {
   proxy_buffers 64 4k;
 
-  proxy_set_header Host $proxy_host;
+  proxy_set_header Host $AEM_ADDR;
```
@hlbraddock @cru-Luis-Rodriguez Should it only set this when proxying to AEM and not WordPress?

cru-Luis-Rodriguez replied:
Yes

Omicron7 replied:
ok, I'll work on that after I get staging ECS multi-container running. It'll be something we need to do in lua.

Omicron7 committed:
AEM Cloud requires X-Forwarded-Host header

Omicron7 committed:
Use thescheme for scheme, make all variables local.

Omicron7 committed:
Only change proxy_host if it's aem cloud.

Omicron7 committed:
Use string.find
